### PR TITLE
Interface for DRM Visibility

### PR DESCRIPTION
**Pull request by kabliz**
- Adds DrmState to ArmadilloState, giving full visibility into DRM status to the client, including the expiration date for content.
- Splits SocketTimeout from HttpResponseCodeException, into ConnectivityException to better differentiate connectivity difficulties from developer error.
- Ensures that ArmadilloState is updated from the main thread consistently.

### CODE REVIEW

File: Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt [hunk 1]
```diff
@@ -261,6 +271,42 @@ internal object Reducer {
...
             }
+            is OpeningLicenseAction -> {
+                val oldDrm = oldState.drmPlaybackState
+                return oldState
+                    .copy(drmPlaybackState = DrmState.LicenseOpening(action.type ?: oldDrm.drmType))
+            }
+            is LicenseAcquiredAction -> {
+                val oldDrm = oldState.drmPlaybackState
+                return oldState
+                    .copy(drmPlaybackState = DrmState.LicenseAcquired(action.type, oldDrm.expireMillis))
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt:282]: For my understanding - what's the difference between `LicenseAcquired` and `LicenseUsable`?

File: Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt [hunk 2]
```diff
@@ -261,6 +271,42 @@ internal object Reducer {
...
+            }
+            is LicenseExpirationDetermined -> {
+                val oldDrm = oldState.drmPlaybackState
+                return oldState
+                    .copy(drmPlaybackState = DrmState.LicenseUsable(oldDrm.drmType, action.expirationMilliseconds))
+            }
+            is LicenseReleasedAction -> {
+                val oldDrm = oldState.drmPlaybackState
+                return oldState
+                    .copy(drmPlaybackState = DrmState.LicenseReleased(oldDrm.drmType,oldDrm.expireMillis, oldDrm.isSessionValid))
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt:297]: What's the difference between `LicenseReleased` and `NoDrm`?

File: Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt [hunk 3]
```diff
@@ -25,7 +27,7 @@ internal interface StateStore {
     }
 }
 
-internal class ArmadilloStateStore(private val reducer: Reducer) :
+internal class ArmadilloStateStore(private val reducer: Reducer, private val appContext: Context) :
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt:30]: Do we have to worry about this leaking the context now that it's holding a reference to it?

Comment (kschults) [reply Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt:30]: Would it help if we held a reference to the looper instead? `ArmadilloStateStore(..., appContext: Context) ... private val looper = appContext.mainLooper`

File: Armadillo/src/main/java/com/scribd/armadillo/di/AppModule.kt [hunk 4]
```diff
@@ -33,7 +33,7 @@ internal class AppModule(private val context: Context) {
     @Singleton
     @PrivateModule
     @Provides
-    fun stateStore(reducer: Reducer): ArmadilloStateStore = ArmadilloStateStore(reducer)
+    fun stateStore(reducer: Reducer): ArmadilloStateStore = ArmadilloStateStore(reducer, context)
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/di/AppModule.kt:36]: Actually, you could even pass the looper from here, since that's what you actually need the context for

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            return DrmSessionManager.DRM_UNSUPPORTED
+        }
+    }
+
+    /** Near identical to DefaultDrmSessionManagerProvider method, except for the indicated line */
+    private fun createManager(drmConfiguration: DrmConfiguration): DrmSessionManager {
+        val dataSourceFactory = this.drmHttpDataSourceFactory
+            ?: DefaultHttpDataSource.Factory().setUserAgent(this.userAgent)
+        val httpDrmCallback = HttpMediaDrmCallback(if (drmConfiguration.licenseUri == null) null else drmConfiguration.licenseUri.toString(), drmConfiguration.forceDefaultLicenseUri, (dataSourceFactory)!!)
+        val var4: UnmodifiableIterator<*> = drmConfiguration.licenseRequestHeaders.entries.iterator()
```
Comment (SeanoAndroid) [inline Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:77]: Maybe clean up the `DefaultDrmSessionManagerProvider` code a bit, or leave it to better convey what it's copied from? idk

Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:77]: This seems like it got eaten a little bit by the class converter. Can you clean this up so it's a little more readable?

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            .setMultiSession(drmConfiguration.multiSession)
+            .setPlayClearSamplesWithoutKeys(drmConfiguration.playClearContentWithoutKey)
+            .setUseDrmSessionsForClearContent(*Ints.toArray(drmConfiguration.forcedSessionTrackTypes))
+            .build(httpDrmCallback)
+        drmSessionManager.setMode(DefaultDrmSessionManager.MODE_PLAYBACK, drmConfiguration.keySetId)
+        return drmSessionManager
+    }
+
+    /** New original provider for this class */
+    class ArmadilloDrmProvider(private val stateStore: StateStore.Modifier) : ExoMediaDrm.Provider {
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:96]: Probably makes more sense for this to be a separate class in its own file than an inner class

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            .setUuidAndExoMediaDrmProvider(drmConfiguration.scheme, ArmadilloDrmProvider(stateStore))
+            .setMultiSession(drmConfiguration.multiSession)
+            .setPlayClearSamplesWithoutKeys(drmConfiguration.playClearContentWithoutKey)
+            .setUseDrmSessionsForClearContent(*Ints.toArray(drmConfiguration.forcedSessionTrackTypes))
+            .build(httpDrmCallback)
+        drmSessionManager.setMode(DefaultDrmSessionManager.MODE_PLAYBACK, drmConfiguration.keySetId)
+        return drmSessionManager
+    }
+
+    /** New original provider for this class */
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:95]: The only thing this does differently is emit events to the dispatcher, right? Can you make that explicit in the docstring

File: Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt [hunk 8]
```diff
@@ -1,36 +1,63 @@
...
         val dataSourceFactory = mediaSourceHelper.createDataSourceFactory(context, request)
 
-        downloadTracker.getDownload(request.url.toUri())?.let {
-            if (it.state != Download.STATE_FAILED) {
-                val mediaItem = drmMediaSourceHelper.createMediaItem(context = context, request = request, isDownload = true)
-                return DownloadHelper.createMediaSource(it.request, dataSourceFactory, drmSessionManagerProvider.get(mediaItem))
-            }
-        }
+        val download = downloadTracker.getDownload(request.url.toUri())
+        val isDownloaded = download != null && download.state != Download.STATE_FAILED
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt:38]: I'm worried about `isDownloaded` here because it's also possible for it to be in progress, which would be misleading. But I don't have a better suggestion. It could be either in progress or finished. Maybe `forDownload`?

File: Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt [hunk 9]
```diff
@@ -1,36 +1,63 @@
...
 
-        val mediaItem = drmMediaSourceHelper.createMediaItem(context = context, request = request, isDownload = false)
-        return DashMediaSource.Factory(dataSourceFactory)
-            .createMediaSource(mediaItem)
+        return if (isDownloaded) {
+            val drmManager = drmSessionManagerProvider.get(mediaItem)
+            DownloadHelper.createMediaSource(download!!.request, dataSourceFactory, drmManager)
+        } else {
+            DashMediaSource.Factory(dataSourceFactory)
+                .setDrmSessionManagerProvider(drmSessionManagerProvider)
```
Comment (kschults) [inline Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt:46]: Should this also only get added if there's DRM, or does it not matter?

File: Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt (hunk 1 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt:282]: 1. Got one from the remote server.
2. It works and playback should be happening.

File: Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt (hunk 2 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt:297]: The content is still loaded and the player hasn't reset for new content yet.    Playback won't happen anymore, but old content states are still present.

File: Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt (hunk 3 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt:30]: This is the app context rather than the activity context, so it shouldn't be a danger for leaks the way `Activity` is

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt (hunk 6 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:96]: I felt that the large amount of single use classes  in separate files made everything very hard to follow.  It was very hard to figure out how to build this feature.

File: Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt (hunk 9 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt:46]: should not matter

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt (hunk 6 above)
Comment (SeanoAndroid) [reply Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:96]: >I felt that the large amount of single use classes in separate files made everything very hard to follow. It was very hard to figure out how to build this feature.

This x1000. This is what I meant by my audio player being a more simplified interpretation of armadillo and could be a good reference of what exactly is being done

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    class ArmadilloDrmProvider(private val stateStore: StateStore.Modifier) : ExoMediaDrm.Provider {
+        private var drmExpirationMillis: Long? = null
+        override fun acquireExoMediaDrm(uuid: UUID): ExoMediaDrm {
+            Log.i(TAG, "drm start")
+            //uses the main MediaDrm object that this class uses originally, then after we add new listeners to it.
+            val instance = FrameworkMediaDrm.newInstance(uuid)
+
+            //ExoMediaDrm.OnEventListener doesn't do anything at all, so its not used
+            if (Util.SDK_INT >= 23) {
+                instance.setOnKeyStatusChangeListener { exoMediaDrm, sessionId, keyStatuses, b ->
```
Comment (rubeus90) [inline Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:105]: If we're gonna keep all the params names (are we using them later?), the last boolean param is actually `hasNewUsableKey` 😁

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt (hunk 5 above)
Comment (rubeus90) [reply Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:77]: Original java code
```java
for (Map.Entry<String, String> entry : drmConfiguration.licenseRequestHeaders.entrySet()) {
      httpDrmCallback.setKeyRequestProperty(entry.getKey(), entry.getValue());
    }

**Review by rubeus90 — APPROVED** [review]

File: Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt (hunk 1 above)
Comment (kschults) [reply Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt:282]: What needs to happen between 1 and 2?

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt (hunk 6 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:96]: Yeah if i were to refactor this project, i would use some Facades, layers or something to help indicate hierarchy and order.

File: Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt (hunk 1 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt:282]: They're different Exoplayer events, - acquisition doesn't necessarily mean that decryption worked.

File: Armadillo/src/main/java/com/scribd/armadillo/di/AppModule.kt (hunk 4 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/di/AppModule.kt:36]: That reads better yea

File: Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt (hunk 3 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/StateStore.kt:30]: Resolved by passing in Handler

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt (hunk 5 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:77]: thank you that helps to see the intended behavior

File: Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt (hunk 8 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt:38]: ah nice catch, this line was there before but i didn't think much of it.  Im changing it to check for Completion, because in progress might not contain the audio that is resuming.

File: Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt [hunk 11]
```diff
@@ -73,11 +71,10 @@ internal class ArmadilloDrmSessionManagerProvider @Inject constructor(private va
     private fun createManager(drmConfiguration: DrmConfiguration): DrmSessionManager {
         val dataSourceFactory = this.drmHttpDataSourceFactory
             ?: DefaultHttpDataSource.Factory().setUserAgent(this.userAgent)
-        val httpDrmCallback = HttpMediaDrmCallback(if (drmConfiguration.licenseUri == null) null else drmConfiguration.licenseUri.toString(), drmConfiguration.forceDefaultLicenseUri, (dataSourceFactory)!!)
-        val var4: UnmodifiableIterator<*> = drmConfiguration.licenseRequestHeaders.entries.iterator()
+        val license = if (drmConfiguration.licenseUri == null) null else  drmConfiguration.licenseUri.toString()
```
Comment (SeanoAndroid) [inline Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt:74]: Any reason not to just use  `drmConfiguration.licenseUri?.toString()`?

**Review by SeanoAndroid — APPROVED** [review]

**Review by kschults — APPROVED** [review]

**Review by SeanoAndroid — APPROVED** [review]

**Review by rubeus90 — APPROVED** [review]
